include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+        encrypt_pdu(pdu, sequence);
+        asio::write(next_layer(), asio::buffer(pdu), error);
+        if (error) {
+            return 0;
+        }
+
+        return asio::buffer_size(sequence);
+    }
+
+    template <typename ConstBufferSequence>
+    auto write_some(const ConstBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = write_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence,
+                          beast::error_code& error) {
+        error = {};
+        if (decrypted_buffer.size() > 0) {
+            // We have already decrypted data ready to be retrieved from the
+            // buffer
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            return consumed;
+        }
+
+        boost::endian::little_int16_t payload_length = 0;
+        size_t transferred = 0;
+        while (transferred < sizeof(payload_length) + payload_length +
+                                 crypto::aead_auth_tag_size) {
+            auto mb = raw_buffer.prepare(max_payload_length);
+            transferred += next_layer().read_some(mb, error);
+            if (error) {
+                return 0;
+            }
+            raw_buffer.commit(transferred);
+            extract_pdu_length(payload_length);
+        }
+
+        auto success = decrypt_pdu(payload_length);
+        if (!success) {
+            // Forged/spoofed message
+            // TODO: use a more appropriate error code
+            error = asio::error::no_permission;
+            decrypted_buffer.consume(payload_length);
+            return 0;
+        }
+
+        auto consumed =
+            boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+        decrypted_buffer.consume(consumed);
+        return consumed;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = read_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
```

Comment:
Several problems here. Observe that your function is known as an **initiating function**. We will refer to the requirements from this page:
http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/reference/asynchronous_operations.html

First you must not call `dispatch`, since that can invoke the handler before returning. The rule for initiating functions is as follows:

_Regardless of whether the asynchronous operation completes immediately or not, the handler will not be invoked from within this function. Invocation of the handler will be performed in a manner equivalent to using boost::asio::io_service::post()._

You should be using `io_service::post` instead of calling dispatch.

The next problem is that you have type-erased the completion handler by wrapping it in a lambda. Instead of calling `post` with the handler type, you are calling it with an anonymous lambda.

Type-erasing the handler erases the allocation, invocation, and continuation hooks. These hooks are described here:
http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/reference/asio_handler_allocate.html
http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/reference/asio_handler_deallocate.html
http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/reference/asio_handler_invoke.html
http://www.boost.org/doc/libs/1_64_0/doc/html/boost_asio/reference/asio_handler_is_continuation.html

The consequence for erasing the hooks is severe; calling your `async_read_some` with a handler that has been wrapped in a strand will result in undefined behavior because the strand wrapper will not be invoked in your call to `io_service::post`, thus breaking invariants.

You need to package the handler in a bind wrapper which preserves the hooks and lets you store additional arguments when invoking the handler. Beast to the rescue :) We will use `beast::bind_handler` to achieve this:
```
next_layer().get_io_service().post(
    beast::bind_handler(
        std::move(init.completion_handler),
        beast::error_code{},
        consumed));
```

I suggest you study the implementation of `bind_handler` to see how it forwards the hooks to the wrapped handler:
https://github.com/boostorg/beast/blob/20a8f7d75b45f0ababc2547f50833b7f4ba636d3/include/boost/beast/core/detail/bind_handler.hpp#L137

I also suggest studying the Composed Operation tutorial in the Beast documentation:
http://www.boost.org/doc/libs/develop/libs/beast/doc/html/beast/using_io/writing_composed_operations.html